flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+implementation operating on `Fortran::evaluate::Expr` must be added to the
+front-end. Otherwise, this step can be skipped.
+
+In each of those file, there is a big `if` statement switching on the
+intrinsic names. Chose the file according to the intrinsic result types. If the
+intrinsic may return several types and would have a similar implementation, the
+implementation should be done in `lib/Evaluate/fold.cpp`.
+Add the name of the intrinsics in the if-chain, and plug the implementation there.
+See other implementations for examples.
+
+Front-end folding operates on `Fortran::evaluate::Expr`, and in particular, with the
+`Fortran::evaluate::Constant` variant of expressions. You will need to be
+familiar with them to do anything. Look at the defining header in `include/Evaluate`
+to get started. 
+
+#### Tests
+To add regression tests for front-end folding, `test_folding.sh` can be used as a test driver.
+It checks that every parameter prefixed by `test_` is folded to `.true.`.
+`test/Evaluate/folding06.f90` is a good example for it.
+
+## Adding intrinsics to the runtime and lowering
+
+It is first needed to decide whether some runtime is needed or not. If the intrinsic can
+be  implemented with a simple sequence of FIR operations, it might be worth to implement it inline.
+Otherwise, if it involves more advanced control flow and dynamic allocations, a runtime implementation
+might be sounder.
+
+### Runtime
+- See: `lib/runtime` folder
+- Where to submit patch: llvm repository through Phabricator.
+
+#### Implementation
+A runtime interface must be designed for the intrinsic.
+The runtime interface is simply the declaration of the runtime function for which lowering must insert calls,
+as well as a small description of what the function is doing (e.g. comments above the declaration).
+Before starting to implement anything, it is best to submit this interface for review. Having the interface
+already submitted can also allow to implement the lowering part in parallel of the runtime implementation.
+See `lib/runtime/character.h` for examples of runtime interfaces.
+
+The design must decide if one or more functions should form the runtime interface.
+It may make sense to have simple entry points for easy cases, and a more complex entry points
+with descriptors to cover all the other cases. When in doubt, it is best to start with a single
+entry point that cover all cases, and leave entry point specialization for later, based on a measurable
+performance gain on actual programs.
+
+The implementation is done in C++, but there is a hard requirement in not depending on
+the C++ runtime library. That is because it is not wished that Fortran runtime depends
+upon libstdc++ or other C++ runtime libraries. This means that templates and classes
+are OK, `#include <c[library]>` should be OK, but any other includes should be avoided
+unless proven they do not bring C++ runtime in.
+
```

Comment:
You may want to note that using a C compiler for linking the test will result in failures if there are C++ runtime libraries needed.

Reply:
I think our "Hello" test actually still links against libc++ because of early bugs in this area.

Reply:
>I think our "Hello" test actually still links against libc++ because of early bugs in this area.

Good point, the runtime was fixed. So it would be worth giving a go removing this from the test. I'll give it a try and make another PR if that works.